File: typescript/packages/core/src/sandbox/remote/driver.ts
import type { LookupResult } from "../../model/protocol";
import type { Sandbox } from "../protocol";

// What a provider package implements to become a threads Sandbox through remoteSandbox(): the
// provider's raw operations over its official SDK. Every method runs inside `within(context)`,
// sends only through a fenced transport, and throws on a transport or provider failure.
// Termination and quiescence are never inferred from inside the guest: only a
// provider control-plane primitive (a whole-sandbox pause or kill) proves them.

export type Sinks = {
  readonly stdout: (chunk: Uint8Array) => void;
  readonly stderr: (chunk: Uint8Array) => void;
};

/** A started process: `exit` resolves once both output streams have ended. */
export type Started = { readonly exit: Promise<number> };

/**
 * How the provider makes a capture quiescent, declared per provider and mode
 * from its documentation, never broader:
 * - paused: a provider-owned pause of the whole sandbox freezes every process for the capture.
 * - stopped: the provider stops the whole sandbox for it; its processes end, so a snapshot is
 *   disruptive to the parent.
 * - unconfirmed: no documented whole-sandbox boundary; the snapshot capability is absent.
 */
export type Quiescence = "paused" | "stopped" | "unconfirmed";

export type Capture = {
  readonly quiescence: Quiescence;
  /**
   * Captures the filesystem into a durable snapshot named by `operationKey` behind the
   * declared boundary, and returns with the sandbox running again.
   */
  readonly take: (
    id: string,
    operationKey: string,
  ) => Promise<{ readonly ref: string; readonly expiresAt: number | null }>;
};

export type Created =
  | { readonly kind: "created"; readonly id: string }
  | {
      readonly kind: "snapshot_missing" | "snapshot_expired";
      readonly message: string;
    };

/** A lookup's answer that never proves absence: `not_found` is only a final lookup's. */
export type NonfinalLookupResult = Exclude<
  LookupResult<string>,
  { readonly status: "not_found" }
>;

/**
 * How a lost create is found by its operation key. Undeclared, a create in flight at the
 * provider may appear later, so `find` can't answer `not_found`. A driver declares
 * `createLookup: "final"` only with a `find` whose absence the provider's contract makes final.
 */
export type Finding =
  | {
      readonly createLookup?: "nonfinal";
      readonly find: (operationKey: string) => Promise<NonfinalLookupResult>;
    }
  | {
      readonly createLookup: "final";
      readonly find: (operationKey: string) => Promise<LookupResult<string>>;
    };

/**
 * How a process key's group is ended. Undeclared, `stopProcess` is best effort and every
 * termination is unknown. A driver declares `termination: "confirmed"` only with a `terminate`
 * that proves its answer, which the kit then uses instead.
 */
export type Stopping = BestEffortStop | ConfirmedTerminate;

export type BestEffortStop = {
  readonly termination?: "unconfirmed";
  /**
   * Best effort: kills the process the provider recorded under `processKey`. It proves
   * nothing: a descendant can outlive it, and the provider's process table runs in the guest.
   */
  readonly stopProcess: (id: string, processKey: string) => Promise<void>;
};

export type ConfirmedTerminate = {
  readonly termination: "confirmed";
  readonly terminate: (
    id: string,
    processKey: string,
  ) => Promise<"terminated" | "already_exited" | "unknown">;
};

export type SandboxDriver = Finding & Stopping & DriverOperations;

type DriverOperations = {
  /**
   * Creates a sandbox that `find(operationKey)` can locate (a tag, label or unique name), from
   * the snapshot `ref` when given. Nothing from the host env is passed into it.
   */
  readonly create: (
    operationKey: string,
    snapshot: string | undefined,
  ) => Promise<Created>;
  /** Whether the sandbox `id` is still live at the provider. */
  readonly exists: (id: string) => Promise<boolean>;
  readonly kill: (id: string) => Promise<"killed" | "already_gone">;
  /**
   * Starts `sh -c script` as root with an empty provider env, recorded by the provider under
   * `processKey` when given. Resolves once the provider accepted it; output reaches `sinks` as
   * it arrives.
   */
  readonly run: (
    id: string,
    script: string,
    sinks: Sinks,
    processKey?: string,
  ) => Promise<Started>;
  readonly write: (id: string, path: string, data: Uint8Array) => Promise<void>;
  readonly read: (id: string, path: string) => Promise<Uint8Array>;
  /** Absent when the provider has no snapshots. */
  readonly snapshot?: Capture;
  readonly deleteSnapshot: (
    ref: string,
  ) => Promise<"released" | "already_gone">;
};

/** When a provider resource dies on its own. null: never (a Docker container doesn't). */
export type ProviderExpiry = {
  readonly sandboxMs: number | null;
  readonly snapshotMs: number | null;
};

/**
 * A provider adapter: the Sandbox protocol plus what it declares beyond SandboxInfo: when its
 * resources expire, and how its snapshots are made quiescent ("none": no snapshots).
 */
export type ProviderSandbox = Sandbox & {
  readonly expiry: ProviderExpiry;
  readonly quiescence: Quiescence | "none";
};

File: python/src/threads/sandbox/remote/driver.py
"""What a provider adapter implements to become a threads `Sandbox` through `RemoteSandbox`
(sandbox.py): the provider's raw operations over its fenced transport. Every call runs inside
`fence.dispatch`, so each request it sends passes the lease fence at its real send point.
A transport or provider failure raises; the driver's `classify` names the expected ones.

Termination and quiescence are never inferred from inside the guest: only a provider
control-plane primitive (a whole-sandbox pause, stop or kill) proves them. So a driver declares
what it can prove, and each stronger declaration carries its proof:
- `FinalLookup`: a `find` whose absence is final. Without it, a lookup never proves absence.
- `Confirmed`: a `terminate` that answers terminated, already_exited or unknown. Without it,
  `stop_process` is best effort and every termination is unknown.
"""

from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import Literal, Protocol

from threads.adapters.sandboxes.fence import Classify
from threads.loop.model import Found, LookupResult, LookupUnknown, NotFoundNonfinal
from threads.loop.tools import Termination
from threads.sandbox.protocol import ExecOutput, SandboxError

type Quiescence = Literal["paused", "stopped", "unconfirmed"]
"""How the provider makes a capture quiescent, declared from its documentation, never broader:
- paused: a provider-owned pause of the whole sandbox freezes every process for the capture.
- stopped: the provider stops the whole sandbox for it; its processes end, so a snapshot is
  disruptive to the parent.
- unconfirmed: no documented whole-sandbox boundary; the snapshot capability is absent."""


class FileError(Exception):
    """The provider refused a file operation for a reason it names (not_found, is_directory,
    permission_denied, too_large, invalid_path)."""

    def __init__(self, error: SandboxError) -> None:
        super().__init__(error.message)
        self.error = error


@dataclass(frozen=True, slots=True)
class Unmade:
    """A create from a snapshot that the provider refused before making anything."""

    code: Literal["snapshot_missing", "snapshot_expired", "snapshot_restore_failed"]
    message: str


@dataclass(frozen=True, slots=True)
class Taken:
    ref: str
    expires_at: int | None


@dataclass(frozen=True, slots=True)
class Capture:
    quiescence: Quiescence
    take: Callable[[str, str], Awaitable[Taken]]
    """(sandbox id, operation key): captures the filesystem into a durable snapshot named by
    the key behind the declared boundary, and returns with the sandbox running again."""
    delete: Callable[[str], Awaitable[Literal["released", "already_gone"]]]
    """Deletes a snapshot by its ref."""


@dataclass(frozen=True, slots=True)
class NonfinalLookup:
    """A create in flight at the provider may appear later, so absence is never proof."""

    find: Callable[[str], Awaitable[Found[str] | NotFoundNonfinal | LookupUnknown]]


@dataclass(frozen=True, slots=True)
class FinalLookup:
    """The provider's contract proves a key it doesn't find was never created."""

    find: Callable[[str], Awaitable[LookupResult[str]]]


@dataclass(frozen=True, slots=True)
class Unconfirmed:
    stop_process: Callable[[str, str], Awaitable[None]]
    """(sandbox id, process key): kills the process the provider recorded under the key. It
    proves nothing: a descendant can outlive it."""


@dataclass(frozen=True, slots=True)
class Confirmed:
    terminate: Callable[[str, str], Awaitable[Termination]]
    """(sandbox id, process key): ends the key's whole process group and says whether it did."""


class SandboxDriver(Protocol):
    @property
    def classify(self) -> Classify:
        """The provider's expected failures; None re-raises (a bug)."""
        ...

    @property
    def lookup(self) -> NonfinalLookup | FinalLookup: ...

    @property
    def termination(self) -> Unconfirmed | Confirmed: ...

    @property
    def capture(self) -> Capture | None:
        """None when the provider has no snapshots."""
        ...

    async def create(self, operation_key: str, snapshot: str | None) -> str | Unmade:
        """The new sandbox's id, made so that `find(operation_key)` locates it (a tag, label
        or unique name), from the snapshot `snapshot` when given, with /workspace ready.
        Nothing from the host env is passed into it."""
        ...

    async def exists(self, sandbox_id: str) -> bool: ...

    async def kill(self, sandbox_id: str) -> Literal["killed", "already_gone"]: ...

    async def run(
        self,
        sandbox_id: str,
        argv: Sequence[str],
        env: Mapping[str, str],
        cwd: str,
        process_key: str | None,
    ) -> ExecOutput:
        """Starts `argv` with `env` through the provider API (never its argv), recorded by the
        provider under `process_key` when given. Returns once the provider accepted it."""
        ...

    async def write(self, sandbox_id: str, path: str, data: bytes) -> None:
        """Raises FileError for a refusal the provider names."""
        ...

    async def read(self, sandbox_id: str, path: str) -> bytes:
        """Raises FileError for a refusal the provider names."""
        ...
